### ISSUE TITLE
Implicit arguments stolen for explicit ones

### ISSUE THREAD

**Issue by treeowl** [idris-lang/Idris-dev#2268]
I was playing around trying to implement the basics of Van Laarhoven lenses, and this weird thing happened:

```idris
Lens : (s,t,a,b:Type) -> Type
Lens s t a b = {f : Type -> Type} -> Functor f -> (a -> f b) -> s -> f t

lfst : Lens (p,q) (r,q) p r
lfst {f=(x,y)} fFunctor pfr = map (\z => (z,y)) (pfr x)
```

The strange thing: `f` is supposed to have type `Type->Type`. Why does it end up being the `s` argument, of type `(p,q)`?

**Comment by treeowl** [comment]
A simpler example:

```idris
(+) {q=S Z} Z
```
happily spits out `1`.

**Comment by treeowl** [comment]
I just realized that these are actually two different sides of a coin—in one case, the problem is in application; in the other, the problem is in abstraction. The solution to both may be the same.

**Comment by treeowl** [comment]
Fixed by 1054d7d.
